Mathlib/Testing/SlimCheck/Sampleable.lean

```diff
@@ -0,0 +1,324 @@
+    exact Nat.zero_le _
+  )
+
+instance Int.sampleable : Sampleable Int where
+  sample := do choose Int (-(←getSize)) (←getSize) (le_trans (Int.neg_nonpos_of_nonneg (Int.ofNat_zero_le _)) (Int.ofNat_zero_le _))
+
+instance Bool.sampleable : Sampleable Bool where
+  sample := chooseAny Bool
+
+/-- This can be specialized into customized `Sampleable Char` instances.
+The resulting instance has `1 / length` chances of making an unrestricted choice of characters
+and it otherwise chooses a character from `chars` with uniform probabilities.  -/
+def Char.sampleable (length : Nat) (chars : List Char) (pos : 0 < chars.length) : Sampleable Char :=
+  {
+    sample := do
+      let x ← choose Nat 0 length (Nat.zero_le _)
+      if x.val == 0 then
+        let n ← Sampleable.sample Nat
+        pure $ Char.ofNat n
+      else
+        elements chars pos
+  }
+
+instance Char.sampleableDefault : Sampleable Char :=
+  Char.sampleable 3 " 0123abcABC:,;`\\/".toList (by decide)
+
+instance Prod.sampleable [Repr α] [Repr β] [Sampleable α] [Sampleable β] : Sampleable (Prod α β) where
+  sample := prodOf (Sampleable.sample α) (Sampleable.sample β)
+
+end Samplers
+
+section Shrinkers
+
+/-- `Nat.shrink' n` creates a list of smaller natural numbers by
+successively dividing `n` by 2 . For example, `Nat.shrink 5 = [2, 1, 0]`. -/
+def Nat.shrink (n : Nat) : List { y : Nat // sizeOfPhantom.sizeOf y < sizeOfPhantom.sizeOf n } :=
+  if h : 0 < n then
+    let m := n/2
+    have h : m < n := by
+      apply Nat.div_lt_self h
+      decide
+    let rest := shrink m
+    let current := ⟨m, by rw [sizeOfPhantom_eq_SizeOf]; exact h⟩
+    current ::
+      rest.map (λ x => {x with property := by
+        let h2 := x.property
+        simp [sizeOfPhantom_eq_SizeOf]
+        simp [sizeOfPhantom_eq_SizeOf] at h2
+        exact Nat.lt_trans h2 h
+      })
+  else
+    []
+
+instance Nat.shrinkable : Shrinkable Nat where
+  shrink := Nat.shrink
+
+/-- `Fin.shrink` works like `Nat.shrink` but instead operates on `Fin`. -/
+def Fin.shrink {n : Nat} (m : Fin n.succ) : List { y : Fin n.succ // sizeOfPhantom.sizeOf y < sizeOfPhantom.sizeOf m } :=
+  let shrinks := Nat.shrink m.val
+  shrinks.map (λ x => {x with property := by
+    have h := x.property
+    simp [sizeOfPhantom_eq_SizeOf]
+    simp [sizeOfPhantom_eq_SizeOf] at h
+    exact Nat.succ_lt_succ $ lt_of_le_of_lt (Nat.mod_le _ _) h})
+
+instance Fin.shrinkable {n : Nat} : Shrinkable (Fin n.succ) where
+  shrink := Fin.shrink
+
+abbrev Int.sizeOfAbs : SizeOf Int := ⟨Int.natAbs⟩
+
+attribute [local instance] Int.sizeOfAbs
+
+/-- `Int.shrink` operates like `Nat.shrink` but also includes the negative variants. -/
+def Int.shrink (n : Int) : List { y : Int // sizeOfPhantom.sizeOf y < sizeOfPhantom.sizeOf n } :=
+  Nat.shrink n.natAbs |>.map λ ⟨x, h⟩ => ⟨-x, (by rw [sizeOfPhantom_eq_SizeOf] at *; simp only [SizeOf.sizeOf]; rw [Int.natAbs_neg]; exact h)⟩
+
+instance Int.shrinkable : Shrinkable Int where
+  shrink := Int.shrink
```

Comment:
```suggestion
/-- `Int.shrinkable` operates like `Nat.shrinkable` but also includes the negative variants. -/
instance Int.shrinkable : Shrinkable Int where
  shrink n :=
    Nat.shrink n.natAbs |>.map λ ⟨x, h⟩ => ⟨-x, (by rw [sizeOfPhantom_eq_SizeOf] at *; simp only [SizeOf.sizeOf]; rw [Int.natAbs_neg]; exact h)⟩
```

I don't think we need to keep the separate definitions if we already have the instance.